### ISSUE TITLE
Proposed API for twin-axis support

### ISSUE THREAD

**Issue by jlstevens** [holoviz/holoviews#5355]
Twin axes is an often requested feature (e.g. see  #396, #1594, #3011) and HoloViews should support this type of plot without having to resort to ugly custom hooks.

In this issue I would like to propose an API for discussion. In particular, I think the most intuitive approach would be to support twin-axes by creating a simple overlay, where one of the components has `yaxis='twin'` specified as a plot option. Right now, the possible values for `yaxis` are `['left', 'right', None, True, False]` which means `'twin'` would be added to this list.

As `'left'` and `'right'` are already supported, using `'twin'` would add axes on the opposite side for the component it is attached to. For instance:

1. `curve1 * curve2.opts(yaxis='twin')`: curve1 has its yaxis on the left (default) and curve2 has its yaxis on the right
2. `curve1.opts(yaxis='right') * curve2.opts(yaxis='twin')`:  curve1 has its yaxis on the right and curve2 has its yaxis on the left

I think this would be intuitive enough given how everything currently works. In principle, this should support overlaying two `NdOverlays` so that you have two legends, one for each side.

Suggested rules:

1. The x-axis dimension must match otherwise trying to use twin-axes should be an error.
2. Twin axes make sense for a number of chart elements (e.g. curves, scatter) but I don't think they really make sense for `Points` and `Paths` which exist in a homogenous 2D space (where you can sensibly imagine rotations). I don't think this option will make sense for all element types (e.g. can this really be sensible for bar charts?)

**Comment by jbednar** [comment]
> Twin axes make sense for a number of chart elements (e.g. curves, scatter) but I don't think they really make sense for Points and Paths which exist in a homogenous 2D space (where you can sensibly imagine rotations). 

I agree that they don't make much sense for Points and Paths, nor for dense rasters or geo maps, but I wouldn't disallow those if it takes extra work to do so. It's hard to anticipate all the ways someone might use this functionality, some of which are abuses but others might end up useful. 

> I don't think this option will make sense for all element types (e.g. can this really be sensible for bar charts?)

It's hard to see overlaying two bar charts as useful since bar charts themselves already support stacking or side by side bars, but again, maybe not worth the effort of figuring out which ones are truly useless, and which ones are just _rarely_ useful.

**Comment by douglas-raillard-arm** [comment]
Haven't put much thoughts in that but an alternative could be a 2nd overlay operator, e.g. the division (maybe it's already taken for something, haven't checked):
```
def __truediv__(self, other):
    self.opts(yaxis='right') * other.opts(yaxis='twin')
```

**Comment by jlstevens** [comment]
That would indeed be a possible approach but we are wary of adding new operators. In particular, we have discussed using `/` to improve the layout system (to define rows and columns more easily).

**Comment by douglas-raillard-arm** [comment]
It's maybe a terrible idea but it just struck me that we actually _can_ have an unlimited supply of infix operators in Python if we are ready to make pylint scream about the missing whitespaces:
```python
class Meta(type):
    def __rlshift__(self, left):
        return self(left)

class Op(metaclass=Meta):
    def __init__(self, left):
        self.left = left

    def __rshift__(self, right):
        return self.op(self.left, right)


class Custom(Op):
    @staticmethod
    def op(left, right):
        return left * right

# We just created a new operator "<<Custom>>"
print(3 <<Custom>> 2)
```

**Comment by csgorham** [comment]
Trying to plot double BoxWhiskers with double y axis for 'height' and 'weight' ... running into so much trouble `def twinyy(plot, element):
    # Setting the second y axis range name and range
    #start, end = (element.range(1))
    label = element.dimensions()[1].pprint_label
    plot.state.extra_y_ranges = {"foo": Range1d(start=4, end=6)}
    # Adding the second axis to the plot. 
    linaxis = LinearAxis(axis_label='Height', y_range_name='foo')
    plot.state.add_layout(linaxis, 'right')
    
    
title = "Weight and Height of Club Members"

boxwhisker1 = hv.BoxWhisker(teams_fifa , 
                            'Club', 'Weight', label=title).opts(show_legend=False, width=600, 
                                                              box_fill_color='Blue', box_line_color='Blue')
boxwhisker2 = hv.BoxWhisker(teams_fifa , 
                           'Club', 'Height', label=title).opts(show_legend=False, width=600, 
                                                              box_fill_color='Red', box_line_color='Red', hooks=[twinyy])

boxwhisker1*boxwhisker2`

**Comment by jlstevens** [comment]
Unfortunately, HoloViews does not support twin axes well right now and any workarounds that might work aren't great. We do want to support it better in the future though!

**Comment by MarcSkovMadsen** [comment]
There is a user requesting it here https://discourse.holoviz.org/t/how-to-implement-secondary-y-axis/4317/2

I also need it for my use cases.

**Comment by jbednar** [comment]
It's a funded task on our list, so I hope we'll get to it soon, but we do have a bunch of priorities to juggle, so it's hard to promise anything!

**Comment by Episkiliski** [comment]
Hi, is there any development on this? 

In my line of work this is crucial, for example to plot Temperature in Primary y-axis (left) and Pressure in Secondary y-axis (right) with Time in the x-axis. This is not allowing me to use HvPlot unfortunately.

Thanks!

**Comment by jlstevens** [comment]
There is! Thank you for bumping this issue!

Sorry, I should have referenced this issue in the PR that was merged (and now released): https://github.com/holoviz/holoviews/pull/5621

The API is to use `multi_y=True` on an overlay and you can find more docs here: https://holoviews.org/user_guide/Customizing_Plots.html#twin-axes

Closing.

**Comment by Episkiliski** [comment]
> There is! Thank you for bumping this issue!
> 
> Sorry, I should have referenced this issue in the PR that was merged (and now released): #5621
> 
> The API is to use `multi_y=True` on an overlay and you can find more docs here: https://holoviews.org/user_guide/Customizing_Plots.html#twin-axes
> 
> Closing.

Thanks, does it have direct application in HvPlot?
